pkg/querysharding/analyzer.go

```diff
@@ -36,7 +70,9 @@ func NewQueryAnalyzer() *QueryAnalyzer {
 //   as shardable by those labels.
 // * otherwise, treat the query as non-shardable.
 // The le label is excluded from sharding.
-func (a *QueryAnalyzer) Analyze(query string) (QueryAnalysis, error) {
+
+func (a *QueryAnalyzer) queryAnalyzer(query string) (QueryAnalysis, error) {
```

Comment:
Let's not add another method for this. Instead we make `Analyze(query string) (QueryAnalysis, error)` an interface and you can implement the `CachedQueryAnalyzer` for this interface which supports LRU cache. Does it make sense?

Reply:
```go
type Analyze interface{
   Analyze(string) (QueryAnalysis , error)
}


type CachedQueryAnalyzer struct {
    cache *lru.Cache , 
}

func (a *CachedQueryAnalyzer) Analyze(query string) (QueryAnalysis , error){
---
}

```
Everything else will remain same. Like this ? 

```go
func (a *CachedQueryAnalyzer) Analyze(query string) (QueryAnalysis, error) {
    if a.cache.Exists(query) { return a.cache.Get(query) }

    // Analyze if needed
    analysis, err := a.Analyze(query)
    
    // Add to cache
    a.cache.Add(query, xxx)
    return ...
}
```
Question : If yes then how would go compiler know which Analyze to call one from CachedQueryAnalyzer or from QueryAnalyzer on 4th line?

Reply:
`analysis, err := a.Analyze(query)` I didn't write the code clearly but it was referring to `a.analyzer.Analyze()`.
So you need to have the original analyzer and have it when create the cached analyzer.

Reply:
```go 
type CachedQueryAnalyzer struct {
    analyzer QueryAnalyzer ,
    cache *lru.Cache , 
}
```
Like this ?

Reply:
Yeah this could work.

Reply:
ok